README.md

```diff
@@ -1 +1,273 @@
+
+> [Implementers] may provide an opt-in for systems where round-trip determinism is a desireable [sic] feature and backward compatibility with old, non-strict data is unnecessary.
+>
+> — [DAG-JSON Spec][DAG-JSON]
+
+As it is critical for guarding against various attacks, the assumptions around canonical encoding MUST be enforced.
+
+## 2.1 Signing CIDs
+
+Rather than validating the inline IPLD, replacing the data with a CID link to the content MAY be used instead. Note while this is very safe (as it impractical to alter a signed hash), this approach mixes data layout with security, and may have a performance, disk, and networking impacts.
+
+### 2.1.1 Caching & Invalidation
+
+Signing CIDs has two additional caching effects
+
+* Signing CIDs enables a simple strategy for caching validation by CID
+* Such a strategy also requires accounting for revocation of the signing keys, and so need to be tagged with this additional metadata
+
+## 2.2 Raw (Noncanonicalized) Data
+
+Canonicalization is not required if data is encoded as raw bytes (multicodec `0x55`). The exact bytes are already present, and MUST not be changed.
+
+# 3 Varsig Format
+
+After being decoded from [unsigned varint]s, a varsig includes the following segments:
+
+```abnf
+varsig = %x34 varsig-content-encoding varsig-header varsig-body
+varsig-content-encoding = unsigned-varint ; The multicodec describing what's being signed over
+varsig-header = unsigned-varint ; Usually the public key code from Multicodec
+varsig-body = *unsigned-varint; Zero or more segments required by the kind of varsig (e.g. raw bytes, hash algorithm, etc)
+```
+
+For example, here is a EdDSA signature for content encoded as DAG-PB:
+
+`0x3470ed01ae3784f03f9ee1163382fa6efa73b0c31ecf58c899c836709303ba4621d1e6df20e09aaa568914290b7ea124f5b38e70b9b69c7de0d216880eac885edd41c302`
+
+### 3.1 Varsig Prefix
+
+The varsig prefix MUST be `0x34`.
+
+### 3.2 Content Multicodec Prefix
+
+The [multicodec] prefix of the encoding scheme used before it's signed over. The default encoding SHOULD be [`0x55` "raw binary"][raw binary multicodec].
+
+Many signature schemes depend on a hash function. Algorithm-sensitive hashing MUST be captured in the Varsig [header algorithm](#3-1-3-signature-header) or [body](#3-1-4-varsig-body), and so MUST NOT be captured in the Content Multicodec Prefix field.
+
+Some examples include:
+
+* `0x55` for raw bytes (no special encoding)
+* `0x70` for DAG-PB
+* `0x0129` for DAG-JSON
+* `0x0202` for [CAR] files
+
+### 3.3 Signature Header
+
+The prefix of the signature algorithm. This is often the [multicodec] of the associated public key, but MAY be unique for the signature type. The code MAY live outside the multicodec table. This field MUST act as a discriminant for how many expected fields come in the varsig body, and what each of them mean.
+
+### 3.4 Varsig Body
+
+The varsig body MUST consist of zero or more segments required by the signature algorithm.
```

Comment:
by `segment` do you mean `<varint-encoded length><bytes>` or a more general "bytes of any kind known by the decoder according to the discriminant"? by which I mean, is every varsig body expressed as concat'd fields of `<varint len><bytes>` or is it unstructured?

Reply:
It's unstructured in general, but is structured by the kind of signature that it is. It always starts with a header: `0x34` followed by a tag for the signature algorithm (e.g. EdDSA: `0xed`). The rest of the structure after that is defined by the header.